source/auth/auth.rst

```diff
@@ -1251,172 +1327,191 @@ mechanism
+The driver MUST ensure that credentials have not changed between when the lock was requested and when it was acquired.  The lock MUST be released
+when the callback call as finished or errored.
+This is because the ``REQUEST_TOKEN_CALLBACK`` may involve human interaction, and refresh tokens might only be able to be used used once.
+
+If the callback does not return an object in the correct form of ``IdpResponse``, the driver MUST raise an error either using the type
+system or by raising an error when non-optional properties are missing .
+The driver MUST NOT attempt to validate the token(s) directly.
+It is expected that if the server changes the expected fields, the SASL exchange will be updated with a version parameter.
+Drivers do not need to attempt to provide old-driver-new-server compatibility.
+
+If no callback is given, the driver MUST must raise a validation error.
 
 Reauthentication
 ````````````````
-When reauthentication is requested by the server (as a 391 error code) and MONGODB-OIDC is in use, the driver MUST
-ensure that the Access Token that was most-recently used to authenticate this connection is not used for subsequent authentication, by marking it as expired. If non-expired Access Token is available in the cache, it should be used as usual.  If a refresh
-callback is given, it will be called as usual.  Otherwise the IdPServerResponse will be cleared if present and authentication will proceed from the request callback.
 
-If the ``sasl`` step(s) fail with a 391 error code and the payload of the command contained ``jwt`` , the driver MUST clear the IdPServerResponse and
-attempt to authenticate one more time starting from
-``PrincipalStepRequest``.  An initial reauthentication may fail for various reasons, such as token expiration or identity provider reconfiguration, so a second reauthentication might be needed.
+When reauthentication is requested by the server (as a 391 error code) and MONGODB-OIDC is in use, the driver MUST perform a reauthentication.
+The driver MUST account for the case of multiple connections hitting a reauthentication error at different times.
+The driver MUST also account for a reauthenication that results from an IdP configuration change on the server.
+To accomplish these goal, the following algorithm is used to handle a reauthenication error:
 
-The driver MUST account for the case of multiple connections hitting a reauthentication error at different times, to prevent unnecessary callback calls.   If another connection has already reauthenticated, then the Access Token should not be expired.  The driver can either cache a token generation id per connection as well as in the main cache, or some other equivalent method to track whether a reauthentication has already occurred.
+- First, see if the Access Token on the MongoClient is different than the Access Token on the connection.
+   - If they are different, optimisitically try to authenticate.  On error, continue.
```

Comment:
```suggestion
   - If they are different, optimisitically try to authenticate.  On error, continue.
```

Can we clear up exactly what this means? Does this mean to use the accessToken from the client and set it on the connection? Does it mean to use the refresh token? Do we need to cache the conversationId from the original authentication flow?

Reply:
Rewrote this whole section to be clearer.